designs/sam-config.md

```diff
@@ -0,0 +1,413 @@
+stack_name="using_config_file"
+capabilities="CAPABILITY_IAM"
+region="us-east-1"
+profile="srirammv"
+```
+
+
+The configuration file can then be potentially intialized
+
+* all sam init projects could come with a sample samconfig file
+
+Showcase configuration values
+-----------------------------
+
+On running SAM CLI commands with `--debug`, SAM CLI can output the values read from the configuration file. This way the user is always informed of the total set of parameters are being used by SAM CLI, when the customers need to debug what parameters are actually being passed to the `sam` commands.
+
+
+Config file in Git Repos
+------------------------
+
+`samconfig.toml` file can be checked into a git repo, so that its ready to use on cloning the repo. if the configuration file does not present all the necesssary parameters, the command fails just as if one had specified the same arguments on the command line directly.
+
+Optionally, if multiple configuration files are checked in. One can change the `SAM_CLI_CONFIG` environment variable to point a different configuration file.
+
+`--env` can also be passed in to deal with custom environments defined in the configuration file.
+
+Error Messages
+---------------
+
+When a custom env is passed in, and such an environment is not found. The error message can highlight all the environments that were found in the given configuration file.
+
+`
+sam build --env devo
+Error: Environment 'devo' was not found in .aws-sam/samconfig.toml , Possible environments are : ['dev', 'prod']
+`
+
+> I'm not in favor the 2nd part of the above error message; not in favor of reporting what TOML table/section prefixes (aka environments) are in the config file. That behavior that is very odd and unfamiliar to me. It also adds additional code + test cases for a feature I don't see value in. The first part of that error message is great. Just tell me that the prefix I specified was not found. Done.
+
+Future
+----------
+
+In the future, based on the file names of the configuration files, the environment could also be inferred.
+
+```
+.aws-sam/samconfig-dev.toml
+.aws-sam/samconfig-beta.toml
+.aws-sam/samconfig-prod.toml
+```
+
+`--env` dev will refer to `.aws-sam/samconfig-dev.toml` and so on.
+
+> I'm not in favor of parsing file names in lieu of TOML table/section prefixes. What are the TOML tables in the first file? Does it contain `[dev.build.parameters]` or just `[build.parameters]` or `[default.build.parameters]`. First is redundant, second is out-of-spec, third is potentially confusing since there could be defaults elsewhere.
+> Instead, I support loading of multiple config files in order they are specified. Similar to how `docker-compose` can have multiple `-f filename` parameters. However, then comes the discussion of merge rules: overwrite/append keys, extend arrays, and application of these merge rules across files across defaults, merge rules for config files of different spec versions, etc.?
```

Comment:
This was infact part of the first implementation and then removed for design on this particular hierachy first. But definitely on the radar and will create another issue for this.